dotcom-rendering/src/web/components/Portal.tsx

```diff
@@ -1,26 +1,31 @@
-import ReactDOM from 'react-dom';
+import { Placeholder } from './Placeholder';
 
-type Props = {
-	rootId: string;
-	children: React.ReactNode;
-};
+interface Props {
+	componentName: string;
+	when?: 'immediate' | 'idle' | 'visible';
+	props?: any;
+	placeholderHeight?: number;
+}
 
-// See the preact portal code for details on how this feature works
-// In particular, we can see that portals can safely be recreated and
-// that will not cause the child component being passed in to be mounted
-// again. Instead it is simply rerendered
-// https://github.com/preactjs/preact/blob/df748d106fb78fbd46d14563b4712f921ccf0300/compat/src/portals.js
-export const Portal = ({ rootId, children }: Props) => {
-	const element = document.getElementById(rootId);
-	if (!element) return null;
-	// First remove any placeholder. Why? Because we sometimes server side render Placeholders in
-	// the root divs where the Portal is being inserted so the page is rendered with the placeholder
-	// showing (to reduce jank). But ReactDOM.createPortal won't replace content so we need to
-	// manually remove it here.
-	const placeholderElement = element.querySelector(
-		'[data-name="placeholder"]',
-	);
-	if (placeholderElement) placeholderElement.remove();
-	const result = ReactDOM.createPortal(children, element);
-	return result;
-};
+/**
+ * Inserts content into the dom client side only
+ *
+ * @param when - When insertion should take place.
+ * 		- immediate - Insert without delay
+ * 		- idle - Insert when browser idle
+ * 		- visible - Insert when the gu-portal element appears in viewport
+ * @param componentName - The name of the component to insert
+ * @param props - An object containing the props for the component
+ * @param placeholderHeight - Height in pixels. If provided, a Placeholder is server rendered
+ *
+ */
+export const Portal = ({
```

Comment:
The _Naming Things is Hard_  question: is it confusing to repurpose the name Portal, which is a [first-class concept in (P)react](https://reactjs.org/docs/portals.html) that does something similar-ish but not really?

Reply:
I guess what this component is doing is marking a place in the server side rendered content and saying, put some html _here_ on the client using these attributes.

I quite like `RenderOnClient` although it is perhaps a bit verbose? What about `Render`? That nicely aligns with the `ReactDOM.hydrate` vs `ReactDOM.render`.

Reply:
Cool, thanks @oliverlloyd, I found this helpful in clarifying the purpose of this component 🙂

You're right that `Render` does align nicely with that React method, assuming people realise that this component name is a reference to that. Personally, I see the word `Render` appearing more and more, and hence meaning less and less! The naming of React's `render` method hasn't scaled well, which is why we see more verbose names like `renderToString` and `renderToStaticMarkup` crop up later. This is why I prefer the more verbose name. 

I'm not sure if that's a _me_ thing, and I would be interested to hear other opinions.

Reply:
Yeah, totally. I'm like you, I have my preference but that's based on my own feelings. `Render` is aligned and concise but it could could also be confusing.

A better way to choose a name should be with wider input and a focus on what will make sense to new people coming to the code with less context.

Aside. It's perhaps interesting to think about how Astro [solved this problem](https://docs.astro.build/core-concepts/component-hydration/index.html#mycomponent-clientonly-). They continued to use their `client` api and incorporated this scenario into that. That wouldn't work for as but maybe there are other examples for similar abstractions we can align with?

Reply:
Thinking this api through some more, I'm starting to think a better approach might be to fold this logic into `Hydrate` and not have `Portal` at all. We simply could ask Hydrate to not always server render it's children, which essentially has the same effect as what this PR is proposing.

The api for `Hydrate` might look something like

```ts
<Hydrate when="immediate" where="client">
  <MyClientOnlyThing />
</Hydrate>
```

or

```ts
<Hydrate when="immediate" ssr={false}>
  <MyClientOnlyThing />
</Hydrate>
```